File: src/nappe/grammar.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from tree_sitter import Language

_ERROR_KINDS: frozenset[str] = frozenset({"ERROR", "MISSING", "UNEXPECTED_TOKEN"})
_PROTECTED_KINDS: frozenset[str] = frozenset({"comment"})
_KEYWORD_KINDS: frozenset[str] = frozenset(
    {
        "def",
        "class",
        "if",
        "elif",
        "else",
        "for",
        "while",
        "try",
        "except",
        "finally",
        "with",
        "return",
        "import",
        "from",
        "as",
        "lambda",
        "yield",
        "assert",
        "del",
        "raise",
        "pass",
        "break",
        "continue",
        "global",
        "nonlocal",
        "async",
        "await",
        "match",
        "case",
        "(",
        ")",
        "[",
        "]",
        "{",
        "}",
        ",",
        ":",
        ";",
        ".",
        "->",
        "=",
        "+=",
        "-=",
        "*=",
        "/=",
        "//=",
        "%=",
        "**=",
        ">>=",
        "<<=",
        "&=",
        "^=",
        "|=",
        "and",
        "or",
        "not",
        "in",
        "is",
        "is not",
        "not in",
    }
)


class Grammar:
    """Wraps a tree-sitter language with helper methods."""

    def __init__(
        self, lang_module: Any, name: str = "python", lang_func: Any = None
    ) -> None:
        if lang_func is not None:
            self._language = Language(lang_func())
        else:
            self._language = Language(lang_module.language())
        self._name = name
        self._field_cache: dict[tuple[str, int], str] = {}
        self._build_field_cache()

    @property
    def language(self) -> Language:
        return self._language

    @property
    def name(self) -> str:
        return self._name

    def is_error_node(self, kind: str) -> bool:
        return kind in _ERROR_KINDS

    def is_protected_node(self, kind: str) -> bool:
        return kind in _PROTECTED_KINDS

    def unwrap_compatible_kinds(self, node_kind: str) -> frozenset[str]:
        return self.subtypes(node_kind)

    def subtypes(self, kind: str) -> frozenset[str]:
        kind_id = self._language.id_for_node_kind(kind, True)
        if kind_id is None:
            return frozenset({kind})
        result: set[str] = {kind}
        worklist = [kind_id]
        seen_ids: set[int] = set()
        while worklist:
            current_id = worklist.pop()
            if current_id in seen_ids:
                continue
            seen_ids.add(current_id)
            subtype_ids = self._language.subtypes(current_id)
            for sid in subtype_ids:
                name = self._language.node_kind_for_id(sid)
                if name is not None and name not in result:
                    result.add(name)
                    worklist.append(sid)
        return frozenset(result)

    def supertypes(self, kind: str) -> frozenset[str]:
        kind_id = self._language.id_for_node_kind(kind, True)
        if kind_id is None:
            return frozenset({kind})
        sup_ids = self._language.supertypes
        result: set[str] = {kind}
        for sid in sup_ids:
            sub_ids = self._language.subtypes(sid)
            for sub_id in sub_ids:
                if sub_id == kind_id:
                    name = self._language.node_kind_for_id(sid)
                    if name is not None:
                        result.add(name)
        return frozenset(result)

    def is_kleene_node(self, node_kind: str, child_kinds: tuple[str, ...]) -> bool:
        named = [k for k in child_kinds if k not in _KEYWORD_KINDS]
        return len(named) >= 2 and len(set(named)) == 1

    def is_subtype(self, child_kind: str, parent_kind: str) -> bool:
        return child_kind in self.subtypes(parent_kind)

    def field_name_for_child(self, parent_kind: str, child_index: int) -> str | None:
        return self._field_cache.get((parent_kind, child_index))

    def _build_field_cache(self) -> None:
        from tree_sitter import Parser

        parser = Parser(self._language)
        sample = (
            b"def f(x, y=1):\n"
            b"    return x + y\n"
            b"if True:\n"
            b"    x = 1\n"
            b"class C:\n"
            b"    def m(self): pass\n"
            b"x = [1]\n"
            b"y = {1: 2}\n"
            b"z = (1,)\n"
            b"for i in x:\n"
            b"    pass\n"
            b"while True:\n"
            b"    break\n"
        )
        tree = parser.parse(sample)

        def _walk(node: object) -> None:
            from tree_sitter import Node as TSNode

            assert isinstance(node, TSNode)
            for i, child in enumerate(node.children):
                fname = node.field_name_for_child(i)
                if fname is not None:
                    self._field_cache[(node.type, i)] = fname
            for child in node.children:
                _walk(child)

        _walk(tree.root_node)


_EXTENSION_MAP: dict[str, str] = {
    ".py": "python",
    ".pyi": "python",
    ".js": "javascript",
    ".mjs": "javascript",
    ".cjs": "javascript",
    ".ts": "typescript",
    ".tsx": "typescript",
    ".rs": "rust",
    ".go": "go",
    ".c": "c",
    ".h": "c",
    ".cpp": "cpp",
    ".cc": "cpp",
    ".cxx": "cpp",
}


def detect_language(path: str) -> str:
    ext = Path(path).suffix
    lang = _EXTENSION_MAP.get(ext)
    if lang is None:
        msg = f"Cannot detect language from extension '{ext}'"
        raise ValueError(msg)
    return lang


def load_grammar(lang_name: str) -> Grammar:
    if lang_name == "python":
        import tree_sitter_python

        return Grammar(tree_sitter_python, name="python")
    if lang_name == "javascript":
        import tree_sitter_javascript

        return Grammar(tree_sitter_javascript, name="javascript")
    if lang_name == "typescript":
        import tree_sitter_typescript

        return Grammar(
            tree_sitter_typescript,
            name="typescript",
            lang_func=tree_sitter_typescript.language_typescript,
        )
    if lang_name == "rust":
        import tree_sitter_rust

        return Grammar(tree_sitter_rust, name="rust")
    if lang_name == "go":
        import tree_sitter_go

        return Grammar(tree_sitter_go, name="go")
    if lang_name == "c":
        import tree_sitter_c

        return Grammar(tree_sitter_c, name="c")
    if lang_name == "cpp":
        import tree_sitter_cpp

        return Grammar(tree_sitter_cpp, name="cpp")
    msg = f"Unsupported language: {lang_name}"
    raise ValueError(msg)

File: src/grammar.rs
use std::collections::{HashMap, HashSet};
use std::path::Path;

pub struct Grammar {
    language: tree_sitter::Language,
    name: String,
    field_cache: HashMap<(String, usize), String>,
}

const ERROR_KINDS: &[&str] = &["ERROR", "MISSING", "UNEXPECTED_TOKEN"];
const PROTECTED_KINDS: &[&str] = &["comment"];
const KEYWORD_KINDS: &[&str] = &[
    "def", "class", "if", "elif", "else", "for", "while", "try", "except", "finally", "with",
    "return", "import", "from", "as", "lambda", "yield", "assert", "del", "raise", "pass", "break",
    "continue", "global", "nonlocal", "async", "await", "match", "case", "(", ")", "[", "]", "{",
    "}", ",", ":", ";", ".", "->", "=", "+=", "-=", "*=", "/=", "//=", "%=", "**=", ">>=", "<<=",
    "&=", "^=", "|=", "and", "or", "not", "in", "is", "is not", "not in",
];

impl Grammar {
    pub fn new(lang_fn: tree_sitter_language::LanguageFn, name: &str) -> Self {
        let language = tree_sitter::Language::new(lang_fn);
        let mut grammar = Grammar {
            language,
            name: name.to_string(),
            field_cache: HashMap::new(),
        };
        grammar.build_field_cache();
        grammar
    }

    pub fn language(&self) -> &tree_sitter::Language {
        &self.language
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn is_error_node(&self, kind: &str) -> bool {
        ERROR_KINDS.contains(&kind)
    }

    pub fn is_protected_node(&self, kind: &str) -> bool {
        PROTECTED_KINDS.contains(&kind)
    }

    pub fn unwrap_compatible_kinds(&self, node_kind: &str) -> HashSet<String> {
        self.subtypes(node_kind)
    }

    pub fn subtypes(&self, kind: &str) -> HashSet<String> {
        let kind_id = self.language.id_for_node_kind(kind, true);
        if kind_id == u16::MAX {
            return HashSet::from([kind.to_string()]);
        }
        let mut result = HashSet::from([kind.to_string()]);
        // Collect all supertypes that contain this kind
        for &sup_id in self.language.supertypes() {
            let sub_ids = self.language.subtypes_for_supertype(sup_id);
            if sub_ids.contains(&kind_id) {
                if let Some(name) = self.language.node_kind_for_id(sup_id) {
                    result.insert(name.to_string());
                }
            }
        }
        result
    }

    pub fn supertypes(&self, kind: &str) -> HashSet<String> {
        let kind_id = self.language.id_for_node_kind(kind, true);
        if kind_id == u16::MAX {
            return HashSet::from([kind.to_string()]);
        }
        let mut result = HashSet::from([kind.to_string()]);
        for &sup_id in self.language.supertypes() {
            let sub_ids = self.language.subtypes_for_supertype(sup_id);
            if sub_ids.contains(&kind_id) {
                if let Some(name) = self.language.node_kind_for_id(sup_id) {
                    result.insert(name.to_string());
                }
            }
        }
        result
    }

    pub fn is_kleene_node(&self, _node_kind: &str, child_kinds: &[String]) -> bool {
        let named: Vec<&str> = child_kinds
            .iter()
            .filter(|k| !KEYWORD_KINDS.contains(&k.as_str()))
            .map(|k| k.as_str())
            .collect();
        named.len() >= 2 && named.iter().collect::<HashSet<_>>().len() == 1
    }

    pub fn is_subtype(&self, child_kind: &str, parent_kind: &str) -> bool {
        self.subtypes(parent_kind).contains(child_kind)
    }

    pub fn field_name_for_child(&self, parent_kind: &str, child_index: usize) -> Option<&str> {
        self.field_cache
            .get(&(parent_kind.to_string(), child_index))
            .map(|s| s.as_str())
    }

    fn build_field_cache(&mut self) {
        let mut parser = tree_sitter::Parser::new();
        parser.set_language(&self.language).unwrap();

        let sample = b"def f(x, y=1):\n\
                        \x20\x20\x20\x20return x + y\n\
                        if True:\n\
                        \x20\x20\x20\x20x = 1\n\
                        class C:\n\
                        \x20\x20\x20\x20def m(self): pass\n\
                        x = [1]\n\
                        y = {1: 2}\n\
                        z = (1,)\n\
                        for i in x:\n\
                        \x20\x20\x20\x20pass\n\
                        while True:\n\
                        \x20\x20\x20\x20break\n";

        if let Some(tree) = parser.parse(sample, None) {
            Self::walk_fields(&tree.root_node(), &mut self.field_cache);
        }
    }

    fn walk_fields(node: &tree_sitter::Node, cache: &mut HashMap<(String, usize), String>) {
        for i in 0..node.child_count() {
            if let Some(fname) = node.field_name_for_child(i as u32) {
                cache.insert((node.kind().to_string(), i), fname.to_string());
            }
        }
        let mut cursor = node.walk();
        for child in node.named_children(&mut cursor) {
            Self::walk_fields(&child, cache);
        }
    }
}

const EXTENSION_MAP: &[(&str, &str)] = &[
    (".py", "python"),
    (".pyi", "python"),
    (".js", "javascript"),
    (".mjs", "javascript"),
    (".cjs", "javascript"),
    (".ts", "typescript"),
    (".tsx", "typescript"),
    (".rs", "rust"),
    (".go", "go"),
    (".c", "c"),
    (".h", "c"),
    (".cpp", "cpp"),
    (".cc", "cpp"),
    (".cxx", "cpp"),
];

pub fn detect_language(path: &str) -> Result<String, String> {
    let ext = Path::new(path)
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| format!(".{}", e))
        .ok_or_else(|| "Cannot detect language from extension".to_string())?;

    EXTENSION_MAP
        .iter()
        .find(|(e, _)| *e == ext)
        .map(|(_, lang)| lang.to_string())
        .ok_or_else(|| format!("Cannot detect language from extension '{}'", ext))
}

pub fn load_grammar(lang_name: &str) -> Result<Grammar, String> {
    match lang_name {
        "python" => Ok(Grammar::new(tree_sitter_python::LANGUAGE, "python")),
        "javascript" => Ok(Grammar::new(tree_sitter_javascript::LANGUAGE, "javascript")),
        "typescript" => Ok(Grammar::new(
            tree_sitter_typescript::LANGUAGE_TYPESCRIPT,
            "typescript",
        )),
        "rust" => Ok(Grammar::new(tree_sitter_rust::LANGUAGE, "rust")),
        "go" => Ok(Grammar::new(tree_sitter_go::LANGUAGE, "go")),
        "c" => Ok(Grammar::new(tree_sitter_c::LANGUAGE, "c")),
        "cpp" => Ok(Grammar::new(tree_sitter_cpp::LANGUAGE, "cpp")),
        _ => Err(format!("Unsupported language: {}", lang_name)),
    }
}
